core/blockchain.go

```diff
@@ -2292,47 +2303,59 @@ func (bc *BlockChain) WriteStakingValidator(v *staking.ValidatorWrapper) error {
 	return nil
 }
 
-// ReadValidatorMap reads the addresses of current all validators
-func (bc *BlockChain) ReadValidatorMap() (map[common.Address]struct{}, error) {
-	if cached, ok := bc.validatorMapCache.Get("validatorMap"); ok {
+// ReadValidatorList reads the addresses of current all validators
+func (bc *BlockChain) ReadValidatorList() ([]common.Address, error) {
+	if cached, ok := bc.validatorListCache.Get("validatorList"); ok {
 		by := cached.([]byte)
-		m := make(map[common.Address]struct{})
+		m := []common.Address{}
 		if err := rlp.DecodeBytes(by, &m); err != nil {
 			return nil, err
 		}
 		return m, nil
 	}
-	return rawdb.ReadValidatorMap(bc.db)
+	return rawdb.ReadValidatorList(bc.db)
 }
 
-// WriteValidatorMap writes the list of validator addresses to database
-func (bc *BlockChain) WriteValidatorMap(addrs map[common.Address]struct{}) error {
-	err := rawdb.WriteValidatorMap(bc.db, addrs)
+// WriteValidatorList writes the list of validator addresses to database
+func (bc *BlockChain) WriteValidatorList(addrs []common.Address) error {
+	err := rawdb.WriteValidatorList(bc.db, addrs)
 	if err != nil {
 		return err
 	}
 	by, err := rlp.EncodeToBytes(addrs)
 	if err != nil {
 		return err
 	}
-	bc.validatorMapCache.Add("validatorMap", by)
+	bc.validatorListCache.Add("validatorList", by)
 	return nil
 }
 
-// UpdateValidatorMap updates the validator map according to staking transaction
-func (bc *BlockChain) UpdateValidatorMap(tx *staking.StakingTransaction) error {
+// UpdateValidatorList updates the validator map according to staking transaction
+func (bc *BlockChain) UpdateValidatorList(tx *staking.StakingTransaction) error {
+	// TODO: simply the logic here in staking/types/transaction.go
+	payload, err := tx.RLPEncodeStakeMsg()
+	if err != nil {
+		return err
+	}
+	decodePayload, err := staking.RLPDecodeStakeMsg(payload, tx.StakingType())
+	if err != nil {
+		return err
+	}
 	switch tx.StakingType() {
 	case staking.DirectiveCreateValidator:
-		createValidator := tx.StakingMessage().(staking.CreateValidator)
-		m, err := bc.ReadValidatorMap()
+		createValidator := decodePayload.(*staking.CreateValidator)
+		list, err := bc.ReadValidatorList()
```

Comment:
Let's add all validators in a single method so we don't have to read ValidatorList once for a single tx? Please add TODO if you don't have time.